hare/eligibility/oracle.go

```diff
@@ -251,56 +240,44 @@ func (o *Oracle) Proof(ctx context.Context, layer types.LayerID, round int32) ([
 
 // Returns a map of all active nodes in the specified layer id
 func (o *Oracle) actives(layer types.LayerID) (map[string]struct{}, error) {
-	sl := roundedSafeLayer(layer, types.LayerID(o.cfg.ConfidenceParam), o.layersPerEpoch, types.LayerID(o.cfg.EpochOffset))
-	safeEp := sl.GetEpoch()
-
-	o.With().Info("safe layer and epoch", sl, safeEp)
-	// check genesis
-	// genesis is for 3 epochs with hare since it can only count active identities found in blocks
-	if safeEp < 3 {
-		return nil, errGenesis
-	}
-
 	// lock until any return
 	// note: no need to lock per safeEp - we do not expect many concurrent requests per safeEp (max two)
 	o.lock.Lock()
 	defer o.lock.Unlock()
 
-	// check cache
-	if val, exist := o.activesCache.Get(safeEp); exist {
-		return val.(map[string]struct{}), nil
+	// we can't read blocks during genesis epochs as there are none
+	if layer.GetEpoch().IsGenesis() {
+		return nil, errGenesis
 	}
 
-	// build a map of all blocks on the current layer
-	mp, err := o.blocksProvider.ContextuallyValidBlock(sl)
-	if err != nil {
-		return nil, err
+	// check cache first
+	if val, exist := o.activesCache.Get(layer.GetEpoch()); exist {
+		return val.(map[string]struct{}), nil
 	}
 
-	// no contextually valid blocks
-	if len(mp) == 0 {
-		o.With().Error("could not calculate hare active set size: no contextually valid blocks",
-			layer,
-			layer.GetEpoch(),
-			log.FieldNamed("safe_layer_id", sl),
-			log.FieldNamed("safe_epoch_id", safeEp))
-		return nil, errNoContextualBlocks
+	// get the active set for the epoch
+	atxs := o.atxdb.GetEpochAtxs(layer.GetEpoch())
+	if len(atxs) == 0 {
+		return nil, fmt.Errorf("empty active set for layer %v in non-genesis epoch %v",
+			layer, layer.GetEpoch())
 	}
 
-	activeMap, err := o.getActiveSet(safeEp-1, mp)
-	if err != nil {
-		o.With().Error("could not retrieve active set size",
-			log.Err(err),
-			layer,
-			layer.GetEpoch(),
-			log.FieldNamed("safe_layer_id", sl),
-			log.FieldNamed("safe_epoch_id", safeEp))
-		return nil, err
+	// extract the nodeIDs
+	activeMap := make(map[string]struct{}, len(atxs))
+	for _, atxid := range atxs {
+		atx, err := o.atxdb.GetAtxHeader(atxid)
+		if err != nil {
+			o.With().Error("failed to fetch atx for layer in non-genesis epoch",
```

Comment:
We have a "standard" error message for cases like this (where we have a reference stored, and then the object itself is missing). I also think that mentioning that it's non-genesis is implied/irrelevant - we'll never try to fetch an ATX for a genesis epoch (since there are none).
```suggestion
			o.With().Error("inconsistent state: failed to get atx header",
```

Reply:
I'm actually wondering if we should log this error here, at all. All consumers of this method log the error if one happens, so this will just be duplicated.

In general, to avoid noisy, redundant logging, I think our policy should be to only log errors where we stop propagating them, unless there's useful context that's not available up the call stack. In this case, the layer is available (and already logged).

Reply:
My strategy is to wrap the errors with slightly more context and then return them. Sometimes it's definitely helpful for debugging to print the error, with all of the local context, before returning it, too.